torch/onnx/experimental/fonnx/v1/export_options.py

```diff
@@ -0,0 +1,13 @@
+@dataclasses.dataclass
+class ExportOptions:
+    """Class for keeping track of ONNX exporter flags can perform."""
+
+    opset_version: Optional[int] = 17
```

Comment:
This feels weird to me.

```python
# top level version 9, inner version 10
dynamo_export(..., opset_version=9, export_options=ExportOptions(opset_version=10))
# top level version 9, inner version None
dynamo_export(..., opset_version=9, export_options=ExportOptions(opset_version=None))
# top level version 9, inner version 17
dynamo_export(..., opset_version=9, export_options=ExportOptions())
# top level version None, inner version None
dynamo_export(..., opset_version=None, export_options=ExportOptions(opset_version=None))
```

They are all legal calls, but I don't know what opset_version we are exporting, for each case.

Reply:
Actually, only the third call is valid. User can set `opset_version` in one place, only. Otherwise, an exception is raised.

We discussed the idea of having arguments on the main API for the most common flags, so that suers don't have to instantiate a full class just to set it. Having "config classed" is not pythonic, but we use it as a backdoor for complex stuff. The docstring should mention that main API arguments take precedence over `ExportOptions` and an exception would be raised if user sets both.

Maybe a better approach would be having an internal `_AllExportOptions` with all internal and external flags while `ExportOptions` only has the public ones, as shown below

```python
@dataclasses.dataclass
class _AllExportOptions:
    """Class for keeping track of ONNX exporter flags can perform."""

    opset_version: Optional[int] = 17
    """opset_version: Target ONNX opset version"""

    dynamic_shape: Optional[bool] = True
    """Hints shape information for input/output tensors.

    When True, all inputs shapes are considered dynamic.
    Otherwise, all inputs shapes are considered static.
    """
@dataclasses.dataclass
class ExportOptions:
    """Class for keeping track of ONNX exporter flags can perform."""

    dynamic_shape: Optional[bool] = True
    """Hints shape information for input/output tensors.

    When True, all inputs shapes are considered dynamic.
    Otherwise, all inputs shapes are considered static.
    """
```

Or another idea would be simply not having have-it-all class with all options. That also makes sense because not all internal components need visibility of all options anyway. so the main `export` API can pass `opset_version` to whatever sub-component when needed, instead of passing `export_options` around.

Reply:
I'm a big proponent of raising an exception here in the case of ambiguity. If there are two non-None values, there is ambiguity that should be called out upon validation.

### Raises due to ambiguity:
> ```python
> # top level version 9, inner version 10
> dynamo_export(..., opset_version=9, export_options=ExportOptions(opset_version=10))
> ```

### Opset version 9 is resolved:
> ```python
> # top level version 9, inner version None
> dynamo_export(..., opset_version=9, export_options=ExportOptions(opset_version=None))
> ```

### Raises due to ambiguity:
> ```python
> # top level version 9, inner version 17
> dynamo_export(..., opset_version=9, export_options=ExportOptions())
> ```

### Default value is implied, but this default is internal and not exposed in API:
> ```python
> # top level version None, inner version None
> dynamo_export(..., opset_version=None, export_options=ExportOptions(opset_version=None))
> ```

Reply:
removed top level arguments, everything must go into `options`